kata04/kata04.go

```diff
@@ -0,0 +1,101 @@
+package kata04
+
+import (
+	"bufio"
+	"errors"
+	"fmt"
+	"log"
+	"math"
+	"os"
+	"strconv"
+	"strings"
+)
+
+func weather() string {
+	maxDiff := 0
+	maxDiffDay := ""
+
+	for i, columns := range getDataFromFile("./weather.dat", []int{0, 1, 2}) {
+		if i == 0 {
+			continue //skip header row
+		}
+		if columns[0] != "mo" {
+			maxTemp, err := strconv.Atoi(columns[1])
+			if err != nil {
+				fmt.Printf("Could not convert %v to int\n", columns[1])
+				continue
+			}
+			minTemp, err := strconv.Atoi(columns[2])
+			if err != nil {
+				fmt.Printf("Could not convert %v to int\n", columns[2])
+				continue
+			}
+			if diff := maxTemp - minTemp; diff > maxDiff {
+				maxDiff = diff
+				maxDiffDay = columns[0]
+			}
+		}
+	}
+	return maxDiffDay
+}
+
+func team() string {
+	minDiff := 100000.0
+	minDiffTeam := ""
+
+	for i, columns := range getDataFromFile("./football.dat", []int{1, 6, 8}) {
+		if i == 0 {
+			continue //skip header row
+		}
+		goalsFor, err := strconv.Atoi(columns[1])
+		if err != nil {
+			fmt.Printf("Could not convert %v to int\n", columns[1])
+			continue
+		}
+		goalsAgainst, err := strconv.Atoi(columns[2])
+		if err != nil {
+			fmt.Printf("Could not convert %v to int\n", columns[2])
+			continue
+		}
+		if diff := math.Abs(float64(goalsFor) - float64(goalsAgainst)); diff < minDiff {
+			minDiff = diff
+			minDiffTeam = columns[0]
+		}
+	}
+	return minDiffTeam
+}
+
+func getDataFromFile(filename string, columns []int) [][]string {
+	file, err := os.Open(filename)
+	if err != nil {
+		log.Fatal(err)
+	}
+	defer file.Close()
+
+	output := make([][]string, 0)
+
+	scanner := bufio.NewScanner(file)
+	for scanner.Scan() {
+		columns, err := getColumns(scanner.Text(), columns, "*")
+		if err == nil {
+			output = append(output, columns)
+		}
+	}
```

Comment:
I poked around a bit for something like pythons `yield` keyword to return an array in pieces as it is built, but i couldn't find anything.  

I seem to be doing this often, and am wondering if there is a better way? 
```
output := initialize variable
for something {
    append to output
}
return output
```

Someone suggested using a channel to return the results as they are created, but that doesn't seem like a better solution due to the overhead of channels.

Reply:
Go tends to be more explicit, and less magical, than interpreted languages like python.  It isn't as explicit as C, but is closer to that than python.
https://go-proverbs.github.io/
"A little copying is better than a little dependency."